File: crates/roughneck-py/tests/typecheck_smoke.py
from __future__ import annotations

from typing import Optional

import roughneck_py as rpy


def log_hook(payload: rpy.HookPayloadDict) -> Optional[rpy.HookDecisionDict]:
    event: rpy.HookEventPayloadName = payload["hook_event_name"]
    if event == "Notification":
        return {"messages": ["typed python hook"]}
    return None


def lookup_release(payload: rpy.JsonValue) -> Optional[rpy.JsonValue]:
    if isinstance(payload, dict) and payload.get("name") == "roughneck":
        return {"version": "0.1.0"}
    return None


config: rpy.DeepAgentConfigDict = {
    "system_prompt": "typed smoke",
    "subagents": {"status": "disabled", "agents": []},
}
agent: rpy.DeepAgent = rpy.create_deep_agent(config)
agent.register_hook("notification", log_hook)
agent.register_tool(
    "lookup_release",
    "Return a canned release version for a package name.",
    {
        "type": "object",
        "properties": {"name": {"type": "string"}},
        "required": ["name"],
    },
    lookup_release,
)
session: rpy.AgentSession = agent.start_session(
    {
        "initial_messages": [
            {"role": "user", "content": "hello"},
        ]
    }
)
response: rpy.SessionInvokeResponseDict = session.invoke(
    {
        "messages": [
            {"role": "user", "content": "world"},
        ]
    }
)
latest = response.get("latest_assistant_message")
if latest is not None:
    role: rpy.RoleName = latest["role"]
    print(role)

File: crates/roughneck-node/tests/typecheck-smoke.ts
import {
  createDeepAgent,
  type DeepAgent,
  type HookDecision,
  type HookPayload,
  type JsonValue,
  type SessionInvokeResponse,
  type ToolSchema,
} from '..'

const hook = (payload: HookPayload): HookDecision | undefined => {
  if (payload.hook_event_name === 'Notification') {
    return { messages: ['typed node hook'] }
  }
  return undefined
}

const releaseToolSchema: ToolSchema = {
  type: 'object',
  properties: {
    name: { type: 'string' },
  },
  required: ['name'],
}

const lookupRelease = (input: JsonValue): JsonValue => {
  if (input && typeof input === 'object' && !Array.isArray(input)) {
    const record = input as Record<string, JsonValue>
    if (record.name === 'roughneck') {
      return { version: '0.1.0' }
    }
  }
  return null
}

const agent: DeepAgent = createDeepAgent({
  system_prompt: 'typed smoke',
  subagents: { status: 'disabled', agents: [] },
})
agent.registerHook('notification', hook)
agent.registerTool(
  'lookup_release',
  'Return a canned release version for a package name.',
  releaseToolSchema,
  lookupRelease,
)

async function main(): Promise<void> {
  const session = await agent.startSession({
    initial_messages: [{ role: 'user', content: 'hello' }],
  })
  const response: SessionInvokeResponse = await session.invoke({
    messages: [{ role: 'user', content: 'world' }],
  })
  const latest = response.latest_assistant_message
  if (latest) {
    const role = latest.role
    console.log(role)
  }
}

void main()
